File: src/offer33.SquenceOfBST/sequence_of_BST.cpp
class Solution1 {
public:
    bool VerifySquenceOfBST(vector<int> sequence) {
        if (sequence.size() == 0) return false;
        return VerifySquenceOfBSTCore(sequence, 0, sequence.size()-1);
    }
    
    bool VerifySquenceOfBSTCore(vector<int>& sequence, int begin, int end){
        if (begin > end)
            return false;
        int i = 0;
        for (i = begin; i < end; i++){
            if (sequence[i] > sequence[end])
                break;
        }
        for (int j = i; j < end; j++){
            if (sequence[j] < sequence[end])
                return false;
        }
        bool left = true;
        if (i > begin+1)
            left = VerifySquenceOfBSTCore(sequence, begin, i-1) ;
        bool right = true;
        if (i < end-1)
            right = VerifySquenceOfBSTCore(sequence, i, end-1);
        return left && right;
    }
};

class Solution2 {
public:
    bool VerifySquenceOfBST(vector<int> sequence) {
        if (sequence.size() == 0) return false;
        int len = sequence.size();
        int i = 0;
        while (len--){
            while (sequence[i++] < sequence[len]);
            while (sequence[i++] > sequence[len]);
            if (i < len)
                return false;
            i = 0;
        }
        return true;
    }
};

File: src/offer33.SquenceOfBST/sequence_of_BST.java
//1
public class Solution {
    public boolean VerifySquenceOfBST(int [] sequence) {
        if(sequence.length == 0){
            return false;
        }
        return VerifyCore(sequence, 0, sequence.length - 1);
    }
    private boolean VerifyCore(int[] sequence, int l, int r){
        if(l > r){
            return false;
        }
        int root = sequence[r];
        int i;
        for(i = 0; i < r; i++){
            if(sequence[i] > root){
                break;
            }
        }
        for(int j = i; j < r; j++){
            if(sequence[j] < root){
                return false;
            }
        }
        boolean left = true;
        if(i-1 > l){
            left = VerifyCore(sequence, l, i-1);
        }
        boolean right = true;
        if(r-1 > i){
            right = VerifyCore(sequence, i, r-1);
        }
        return left && right;
    }
}
//2
public class Solution {
    public boolean VerifySquenceOfBST(int [] sequence) {
        if(sequence.length == 0){
            return false;
        }
        int len = sequence.length;
        while((--len) > 0){
            int i = 0;
            while(sequence[i] < sequence[len]){
                i++;
            }
            while(sequence[i] > sequence[len]){
                i++;
            }
            if(i < len){
                return false;
            }
        }
        return true;
    }
}